### ISSUE TITLE
MinGW failures - due to symlink

### ISSUE THREAD

**Issue by MSP-Greg** [ruby/spec#445]
@eregon 

Good day,  as you know, I'm the crazy MinGW guy.  At present, using `make test-spec` has [two failures](https://msp-greg.github.io/file.mingw_test-all.html#test-spec-2-failures).   Running `mspec` has no failures.

I believe you touched on the issue of varying build systems in a ruby-lang issue.  Well, here's another.  I've mostly been building 2.5 trunk, but for various reasons, I've built other versions.  So, I've got a git repo of ruby, and that is the source for my build system.

Since my system is loosely based on the MSYS2 / MinGW packaging system, I have a symlink for `src` within the standard build system folders to the ruby git repo folder.

Hence, the two failures, where `realdirpath` and `realpath` do not match `__FILE__`.

Would you accept a patch that allows the two tests to determine if `__FILE__` is a symlink?

A method like the following does so:

```ruby
fn = 'E:/GitHub/ruby-loco/src/ruby/spec/rubyspec/core/file/realdirpath_spec.rb'

real = File.realdirpath(fn)
#real = File.realpath(fn)

def symlink_in_path?(fn)
  fn_strip = fn.dup
  found_symlink = false
  while fn_temp = File.dirname(fn_strip) do
    found_symlink = File.symlink?(fn_temp)
    break if found_symlink || fn_temp == fn_strip  # reached last dir or drive
    fn_strip = fn_temp
  end
  found_symlink
end

if real != fn
  puts symlink_in_path?(fn)
end
```

On Windows, one has to walk up the path to check for whether `File.symlink?` is true...

**Comment by eregon** [comment]
Yes, of course any improvement ro ruby/spec is welcome, especially if it fixes the specs on some platform/setup.

I am unsure, does File.symlink? and File.realpath work on Windows at all? Or just on mingw builds?

I think an easy way to solve this is to not use `__FILE__` but a simple file in rubyspec_temp, like the spec below, except the file would exist:
https://github.com/ruby/spec/blob/ec1c161a8341a1ceb4075484e8a630be5fecada1/core/file/realdirpath_spec.rb#L97

But, if symlinks work on mingw, we should also test them similarly to how we do on other platforms.
Do you know the semantics of symlink-related Ruby methods on Windows? Is it documented somewhere?

**Comment by MSP-Greg** [comment]
@eregon 

Getting a bit ahead of things.

> I am unsure, does File.symlink? and File.realpath work on Windows at all? Or just on mingw builds?

I'll check, but it will be a bit.  I've got a mswin build from [unak / mswin-build](https://github.com/unak/mswin-build), I'll check with that.  I suspect both work.

On Windows, I believe `symlink?` and `real*` work fine,  The issue is creating symlinks, although I haven't checked all types.  On Windows, creating some types of symlinks require elevated privileges, which are enforced on most user systems, but can be disabled on servers.

So, I don't think there's anyway to test `symlink?` and `real*`, since there's no way for the test system to create a symlink.

Hence, I think using `__FILE__` is fine, but a temp file would be easier, since we'd know that `real*` would be equal to the regular name.

**Comment by eregon** [comment]
https://github.com/ruby/ruby/commit/6f199eb803e2b3c75521a0aac95909d86923a916 should fix both failures (it's committed in ruby/ruby).

So essentially `File.symlink` raises an error on Windows user systems? What exception is raised?

**Comment by MSP-Greg** [comment]
@eregon 

```
irb(main):001:0> File.symlink("E:/temp", "E:/Temp2")
Errno::EACCES: Permission denied @ rb_file_s_symlink - (E:/temp, E:/Temp2)
        from (irb):1:in `symlink'
        from (irb):1
        from C:/ruby25_64/bin/irb.cmd:19:in `<main>'
```

**Comment by MSP-Greg** [comment]
@eregon 

Just checked the commit, both tests pass (running none parallel).

My solution was a bit complicated.  I guess I really don't like temp files in test code.  Problems in `readline` tests in `test-all` also...

Feel free to close.
